sys/vm/vm_cheri_revoke.c

```diff
@@ -679,27 +690,44 @@ vm_cheri_revoke_object_at(const struct vm_cheri_revoke_cookie *crc, int flags,
 		case PMAP_CAPLOADGEN_ALREADY:
 			break;
 
+		case PMAP_CAPLOADGEN_SCAN_RO_WIRED:
+			vm_page_unwire_in_situ(m2);
+			break;
+
+		case PMAP_CAPLOADGEN_SCAN_RW_XBUSIED:
+		case PMAP_CAPLOADGEN_SCAN_RO_XBUSIED:
+			vm_page_xunbusy(m2);
+			break;
+
+		case PMAP_CAPLOADGEN_TEARDOWN:
+			break;
+
 		default:
 			panic("Bad return from didvm caploadgen update: %d",
 			    pres);
 		}
 
-		/* pmap_caploadgen_page will have unwired/unbusied for us */
-		KASSERT(m2 == NULL, ("LS didvm upd !NULL?"));
+		/*
+		 * If the page wasn't busy, it's possible that its mapping was
+		 * replaced while servicing a page fault.  In this case,
+		 * however, the new page should have been scanned by the page
+		 * fault handler.
+		 */
+		KASSERT(m2 == NULL || !mxbusy,
+		    ("mapping of %p was replaced by a mapping of %p", m, m2));
+
+		/* pmap_caploadgen_update() will have released the page. */
 		mwired = false;
 		mxbusy = false;
```

Comment:
Given these are used outside INVARIANTS code is it correct to have these set inside an #ifdef INVARIANTS block? That seems highly suspect.

Reply:
Ack, sounds good to me

Reply:
Sounds reasonable.  I think the current interface came about before it was clear that any replacement mappings that happened between the two `caploadgen_update` calls was guaranteed to have DTRT re: revocation, so there was some idea of a "no, try sweeping this one instead" to the interface that has long since gone away.